### ISSUE TITLE
piping stdin into kubectl exec sometimes hangs with docker 1.13+

### ISSUE THREAD

**Issue by liggitt** [kubernetes/kubernetes#61557]
the following will frequently hang:
```
# start "mypod" paused with a container containing the "cat" command
echo -n a | kubectl exec -i mypod cat
echo -n a | kubectl exec -i mypod cat
echo -n a | kubectl exec -i mypod cat
echo -n a | kubectl exec -i mypod cat
```

The issue is that the docker client does not reliably close the input stream.

Tracked upstream in https://github.com/moby/moby/issues/36516

Fixed in https://github.com/moby/moby/pull/36517

This issue will persist until kubernetes picks up a version of the docker client containing that fix.

**Comment by tuminoid** [comment]
Hitting this one with RHEL's docker 1.13.1. Workaround for us is to downgrade Docker to 1.12.6.

**Comment by tuminoid** [comment]
This also occurs with `kubectl cp`. Tested up to k8s `1.11.2`.

Any idea when k8s might pick up fixed Docker client? This is super-annoying bug, that makes scripting with `kubectl` very frustrating, and I'm wondering why it doesn't bother more people.

**Comment by tuminoid** [comment]
@liggitt, any updates? k8s 1.14 has Docker 1.13 as minimum supported Docker version, meaning we cannot stick to 1.12.6 much longer to workaround this. Having `exec` and more importantly `cp` hang randomly is PITA.

**Comment by tuminoid** [comment]
This appears to be fixed in 1.13.x, but it wasn't mentioned in any changelog. Can someone link the PR this was fixed in?

**Comment by liggitt** [comment]
fixed in https://github.com/kubernetes/kubernetes/commit/31ffd9f88130c1c0cbb991f90b49887327567398 in 1.12+

/close
